### PR TITLE
working on new constructor

### PR DESCRIPTION
**Pull request by JensWehner**
Cudamatrix can now be created from a block of a matrix, which makes some things easier.

### CODE REVIEW

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@votca-bot changelog: added a more general constructor to cudamatrix

**Review by felipeZ — COMMENTED** [review]
I really like the idea to extend the interface to blocks. I am just curious to know whether sending the blocks back and for is still more efficient than computing then in the CPU.

**Comment by JensWehner** [comment]
yes, good question. I do not know, I assume that the cuda library can do it more efficiently than I can.

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@votca-bot format

File: src/libxtp/cudamatrix.cc [hunk 1]
```diff
@@ -29,21 +29,45 @@ void checkCuda(cudaError_t result) {
...
+    case CUBLAS_STATUS_MAPPING_ERROR:
+      return "CUBLAS_STATUS_MAPPING_ERROR";
+    case CUBLAS_STATUS_EXECUTION_FAILED:
+      return "CUBLAS_STATUS_EXECUTION_FAILED";
+    case CUBLAS_STATUS_INTERNAL_ERROR:
+      return "CUBLAS_STATUS_INTERNAL_ERROR";
+    case CUBLAS_STATUS_NOT_SUPPORTED:
+      return "CUBLAS_STATUS_NOT_SUPPORTED";
+    case CUBLAS_STATUS_LICENSE_ERROR:
+      return "CUBLAS_STATUS_LICENSE_ERROR";
```
Comment (junghans) [inline src/libxtp/cudamatrix.cc:60]: Is there a smart way to do this? Seems like this is doomed to get out of sync immediately.

Comment (JensWehner) [reply src/libxtp/cudamatrix.cc:60]: at the moment there is not, cublas does not have a function for that.

File: src/libxtp/openmp_cuda.cc [hunk 2]
```diff
@@ -68,8 +63,26 @@ void OpenMP_CUDA::setOperators(const std::vector<Eigen::MatrixXd>&,
...
+#endif
+
+Index OpenMP_CUDA::getParentThreadId() const {
+  return inside_Parallel_region_ ? threadID_parent_ : OPENMP::getThreadId();
+}
+Index OpenMP_CUDA::getLocalThreadId(Index ParentThreadId) const {
+  return inside_Parallel_region_ ? 0 : ParentThreadId;
+}
+Index OpenMP_CUDA::getNumberThreads() const {
+  return inside_Parallel_region_ ? 1 : OPENMP::getMaxThreads();
```
Comment (felipeZ) [inline src/libxtp/openmp_cuda.cc:85]: This new functionality to manage the OpenMP/GPU threads improves significantly the code's readability. I am just wondering if in some point there should be some graph/schema explaining visually how the different parts interact. The code is getting more robust but there are significant more abstraction that is mostly in your head. The code quality is pretty good, I am just saying that maybe a graph/picture/schema may help to get the overall picture.

File: src/libxtp/openmp_cuda.cc [hunk 3]
```diff
@@ -92,18 +105,13 @@ void OpenMP_CUDA::MultiplyRight(Eigen::MatrixXd& tensor) {
...
-
-  Index threadid =
-      inside_Parallel_region_ ? threadID_parent_ : OPENMP::getThreadId();
-  if (isInVector(threadid, gpuIDs_)) {
-    if (inside_Parallel_region_) {
-      threadid = 0;
-    }
-    checkCuda(cudaSetDevice(cuda_pips_[threadid]->getDeviceId()));
-    temp_[threadid].A->copy_to_gpu(tensor);
-    cuda_pips_[threadid]->gemm(*temp_[threadid].A, *temp_[threadid].B,
```
Comment (felipeZ) [inline src/libxtp/openmp_cuda.cc:104]: This is so much better than using pointer notation!

**Review by felipeZ — APPROVED** [review]
I like more and more the new constructors and all the functionality to abstract away the GPU interactions.

File: src/libxtp/openmp_cuda.cc (hunk 2 above)
Comment (JensWehner) [reply src/libxtp/openmp_cuda.cc:85]: Good point, this class will get uglier again unfortunately.

**Pull request merged**